tgui/packages/tgui/components/Dropdown.tsx

```diff
@@ -0,0 +1,297 @@
+import { createPopper, VirtualElement } from '@popperjs/core';
+import { classes } from 'common/react';
+import { Component, findDOMfromVNode, InfernoNode, render } from 'inferno';
+import { Box, BoxProps } from './Box';
+import { Icon } from './Icon';
+
+export interface DropdownEntry {
+  displayText: string | number | InfernoNode;
+  value: string | number | Enumerator;
+}
+
+type DropdownUniqueProps = {
+  options: string[] | DropdownEntry[];
+  icon?: string;
+  iconRotation?: number;
+  clipSelectedText?: boolean;
+  width?: string;
+  menuWidth?: string;
+  over?: boolean;
+  color?: string;
+  nochevron?: boolean;
+  displayText?: string | number | InfernoNode;
+  onClick?: (event) => void;
+  // you freaks really are just doing anything with this shit
+  selected?: any;
+  onSelected?: (selected: any) => void;
+};
+
+export type DropdownProps = BoxProps & DropdownUniqueProps;
+
+const DEFAULT_OPTIONS = {
+  placement: 'left-start',
+  modifiers: [
+    {
+      name: 'eventListeners',
+      enabled: false,
+    },
+  ],
+};
+const NULL_RECT: DOMRect = {
+  width: 0,
+  height: 0,
+  top: 0,
+  right: 0,
+  bottom: 0,
+  left: 0,
+  x: 0,
+  y: 0,
+  toJSON: () => null,
+};
+
+type DropdownState = {
+  selected?: string;
+  open: boolean;
+};
+
+const DROPDOWN_DEFAULT_CLASSNAMES = 'Layout Dropdown__menu';
+const DROPDOWN_SCROLL_CLASSNAMES = 'Layout Dropdown__menu-scroll';
+
+export class Dropdown extends Component<DropdownProps, DropdownState> {
+  static renderedMenu: HTMLDivElement | undefined;
+  static singletonPopper: ReturnType<typeof createPopper> | undefined;
+  static currentOpenMenu: Element | undefined;
+  static virtualElement: VirtualElement = {
+    getBoundingClientRect: () =>
+      Dropdown.currentOpenMenu?.getBoundingClientRect() ?? NULL_RECT,
+  };
+  menuContents: any;
+  handleClick: any;
+  state: DropdownState = {
+    open: false,
+  };
+
+  constructor() {
+    super();
+
+    this.handleClick = () => {
+      if (this.state.open) {
+        this.setOpen(false);
+      }
+    };
+  }
+
+  getDOMNode() {
+    return findDOMfromVNode(this.$LI, true);
+  }
+
+  componentDidMount() {
+    const domNode = this.getDOMNode();
+
+    if (!domNode) {
+      return;
+    }
+  }
+
+  openMenu() {
+    let renderedMenu = Dropdown.renderedMenu;
+    if (renderedMenu === undefined) {
+      renderedMenu = document.createElement('div');
+      renderedMenu.className = DROPDOWN_DEFAULT_CLASSNAMES;
+      document.body.appendChild(renderedMenu);
+      Dropdown.renderedMenu = renderedMenu;
+    }
+
+    const domNode = this.getDOMNode()!;
+    Dropdown.currentOpenMenu = domNode;
+
+    renderedMenu.scrollTop = 0;
+    renderedMenu.style.width =
+      // Hack, but domNode should *always* be the parent control meaning it will have width
+      // @ts-ignore
```

Comment:
Really wish ts did not get disabled, despite this- is it possible to use an `in` here? `"menuWidth" in this.props`

Reply:
the ts disable is for for the `domNode.offsetWidth` line, the issue is that `domNode` is a var acquired via an already very hacky method and there isn't a safe way to cast it to specify.